File: esn_tutorial/src/limit_switch.cpp
#include <esn_tutorial/limit_switch.hpp>

namespace esn{

LimitSwitch::LimitSwitch() : Node("limit_switch_node")
{
    // Leggi i parametri
    this->declare_parameter("topic_name", "object_arrived");
    std::string topic_name = this->get_parameter("topic_name").as_string();

    this->declare_parameter("delta_time", 10.0);
    this->delta_t_ = this->get_parameter("delta_time").as_double();

    this->publisher_ = this->create_publisher<std_msgs::msg::Bool>(topic_name, 10); //10 è la dimensione della coda per la pubblicazione dei messaggi
    this->last_publish_time_ = this->get_clock()->now();

    this->timer_ = this->create_wall_timer(
        std::chrono::milliseconds(1000),
        std::bind(&LimitSwitch::timer_callback, this)
        );

    RCLCPP_INFO_STREAM(this->get_logger(),"The limit switch will be triggered every "<<this->delta_t_<<" seconds");
}

void LimitSwitch::timer_callback()
{
    rclcpp::Time now = this->get_clock()->now();

    std_msgs::msg::Bool msg;
    if((now - last_publish_time_) > rclcpp::Duration::from_seconds(this->delta_t_))
    {
        msg.data = true;
        last_publish_time_ = now;
    }
    else
    {
        msg.data = false;
    }

    this->publisher_->publish(msg);
}
}

File: esn_tutorial_py/esn_tutorial_py/limit_switch.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import Bool
from rclpy.duration import Duration


class LimitSwitch(Node):
    """Nodo ROS2 che pubblica un Bool ogni secondo, True ogni delta_t secondi."""

    def __init__(self):
        super().__init__('limit_switch_node')

        # Dichiarazione e lettura dei parametri
        self.declare_parameter('topic_name', '/object_arrived')
        topic_name = self.get_parameter('topic_name').get_parameter_value().string_value

        self.declare_parameter('delta_time', 10.0)
        self._delta_t = self.get_parameter('delta_time').get_parameter_value().double_value

        # Publisher
        self._publisher = self.create_publisher(Bool, topic_name, 10)
        
        # Tempo dell'ultima pubblicazione "True"
        self._last_publish_time = self.get_clock().now()
        
        self._timer = self.create_timer(1.0, self._timer_callback)

        self.get_logger().info(
            f"LimitSwitch node started. Publishing on '{topic_name}' every 1s "
            f"(True ogni {self._delta_t} secondi)"
        )

    def _timer_callback(self):
        now = self.get_clock().now()
        msg = Bool()

        if (now - self._last_publish_time) > Duration(seconds=self._delta_t):
            msg.data = True
            self._last_publish_time = now
        else:
            msg.data = False

        self._publisher.publish(msg)
        self.get_logger().debug(f"Published: {msg.data}")